src/ray/raylet/scheduling/cluster_resource_manager.cc

```diff
@@ -0,0 +1,261 @@
+#include "ray/common/ray_config.h"
+#include "ray/util/container_util.h"
+
+namespace ray {
+
+ClusterResourceManager::ClusterResourceManager(StringIdMap &string_to_int_map)
+    : nodes_{}, string_to_int_map_{string_to_int_map} {}
+
+void ClusterResourceManager::AddOrUpdateNode(
+    const std::string &node_id,
+    const absl::flat_hash_map<std::string, double> &resources_total,
+    const absl::flat_hash_map<std::string, double> &resources_available) {
+  NodeResources node_resources = ResourceMapToNodeResources(
+      string_to_int_map_, resources_total, resources_available);
+  AddOrUpdateNode(string_to_int_map_.Insert(node_id), node_resources);
+}
+
+void ClusterResourceManager::AddOrUpdateNode(int64_t node_id,
+                                             const NodeResources &node_resources) {
+  RAY_LOG(DEBUG) << "Update node info, node_id: " << node_id << ", node_resources: "
+                 << node_resources.DebugString(string_to_int_map_);
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    // This node is new, so add it to the map.
+    nodes_.emplace(node_id, node_resources);
+  } else {
+    // This node exists, so update its resources.
+    it->second = Node(node_resources);
+  }
+}
+
+bool ClusterResourceManager::UpdateNode(const std::string &node_id_string,
+                                        const rpc::ResourcesData &resource_data) {
+  auto node_id = string_to_int_map_.Insert(node_id_string);
+  if (!nodes_.contains(node_id)) {
+    return false;
+  }
+
+  auto resources_total = MapFromProtobuf(resource_data.resources_total());
+  auto resources_available = MapFromProtobuf(resource_data.resources_available());
+  NodeResources node_resources = ResourceMapToNodeResources(
+      string_to_int_map_, resources_total, resources_available);
+  NodeResources local_view;
+  RAY_CHECK(GetNodeResources(node_id, &local_view));
+
+  if (resource_data.resources_total_size() > 0) {
+    for (size_t i = 0; i < node_resources.predefined_resources.size(); ++i) {
+      local_view.predefined_resources[i].total =
+          node_resources.predefined_resources[i].total;
+    }
+    for (auto &entry : node_resources.custom_resources) {
+      local_view.custom_resources[entry.first].total = entry.second.total;
+    }
+  }
+
+  if (resource_data.resources_available_changed()) {
+    for (size_t i = 0; i < node_resources.predefined_resources.size(); ++i) {
+      local_view.predefined_resources[i].available =
+          node_resources.predefined_resources[i].available;
+    }
+    for (auto &entry : node_resources.custom_resources) {
+      local_view.custom_resources[entry.first].available = entry.second.available;
+    }
+
+    local_view.object_pulls_queued = resource_data.object_pulls_queued();
+  }
+
+  AddOrUpdateNode(node_id, local_view);
+  return true;
+}
+
+bool ClusterResourceManager::RemoveNode(int64_t node_id) {
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    // Node not found.
```

Comment:
nit: is it worth printing some logs when this happens? It seems like an error shouldn't happen frequently.